File: kata-solutions/gilded-rose-kata/gilded-rose-python/2-lift-up-conditional/src/gilded_rose.py
class GildedRose:
    def __init__(self, items):
        self.items = items

    def update_item_quality(self, item):
        if item.name == "Aged Brie":
            if item.quality < 50:
                item.quality = item.quality + 1
            item.sell_in = item.sell_in - 1
            if item.sell_in < 0:
                if item.quality < 50:
                    item.quality = item.quality + 1
        elif item.name == "Backstage passes to a TAFKAL80ETC concert":
            if item.quality < 50:
                item.quality = item.quality + 1
                if item.sell_in < 11:
                    if item.quality < 50:
                        item.quality = item.quality + 1
                if item.sell_in < 6:
                    if item.quality < 50:
                        item.quality = item.quality + 1

            item.sell_in = item.sell_in - 1
            if item.sell_in < 0:
                item.quality = item.quality - item.quality
        elif item.name == "Sulfuras, Hand of Ragnaros":
            pass
        else:
            if item.quality > 0:
                item.quality = item.quality - 1
            item.sell_in = item.sell_in - 1

    def update_quality(self):
        for item in self.items:
            self.update_item_quality(item)


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)

File: kata-solutions/gilded-rose-kata/gilded-rose-typescript/4-gilded-rose-extended/src/gilded-rose.ts
'use strict';

import internal from "stream";

export class Item {
  name: string = "";
  sellIn: number = 0;
  quality: number = 0;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }

  public incrementQuality() {
    if (this.quality < 50)
      this.quality = this.quality + 1;
  }

  public decrementQuality() {
    if (this.quality > 0)
      this.quality = this.quality - 1;
  }

  public updateToday() {
    this.decrementQuality();
  }

  public updateTomorrow() {
    this.decrementQuality();
  }

  public update() {
    this.updateToday();
    this.sellIn = this.sellIn - 1;
    this.updateTomorrow();
  }
}

export class AgedBrie extends Item {
  constructor(sellIn: number, quality: number) {
    super("Aged Brie", sellIn, quality);
  }

  public updateToday() {
    this.incrementQuality();
  }

  public updateTomorrow() {
    this.incrementQuality();
  }
}

export class Sulfuras extends Item {
  constructor(sellIn: number, quality: number) {
    super("Sulfuras, Hand of Ragnaros", sellIn, quality);
  }

  public update() {
  }
}

export class ConjuredItem extends Item {
  constructor(name: string, sellIn: number, quality: number) {
    super(name, sellIn, quality);
  }

  public updateToday() {
    if (this.sellIn == 0) {
      this.decrementQuality();
      this.decrementQuality();
    }

    if (this.sellIn == 5)
      this.decrementQuality();

    this.decrementQuality();
  }

  public updateTomorrow(): void {
    this.decrementQuality();
  }
}

export class BackstagePass extends Item {
  constructor(sellIn: number, quality: number) {
    super("Backstage passes to a TAFKAL80ETC concert", sellIn, quality);
  }

  public updateToday() {
    this.incrementQuality()
    if (this.sellIn < 11)
      this.incrementQuality()
    if (this.sellIn < 6)
      this.incrementQuality()
  }

  public updateTomorrow() {
    if (this.sellIn < 0)
      this.quality = 0;
  }
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  public updateQuality(): Array<Item> {
    this.items.forEach(item => item.update());
    return this.items;
  }
}
